src/com/xilinx/rapidwright/rwroute/RouteNodeGraph.java

```diff
@@ -175,69 +182,115 @@ protected RouteNodeGraph(Design design, RWRouteConfig config, Map<Tile, RouteNod
+                    if (baseIntentCode == IntentCode.NODE_CLE_BNODE || baseIntentCode == IntentCode.NODE_CLE_CNODE) {
+                        ew = ew.equals("E") ? "W" : "E";
                     }
-                } else if (m.group(6) != null) {
-                    int i = Integer.valueOf(m.group(6));
-                    if (i < 32 || (isUltraScale && (i >= 64 && i < 96))) {
+                    if (ew.equals("E")) {
                         eastWires.set(baseNode.getWireIndex());
                     } else {
-                        assert(i < 64 || (isUltraScale && (i >= 96 && i < 128)));
+                        assert(ew.equals("W"));
                         westWires.set(baseNode.getWireIndex());
                     }
+                } else {
+                    if ((inode = m.group("inode")) != null) {
+                        int i = Integer.valueOf(inode);
+                        if (i < 32 || ((isUltraScale || isVersal) && i >= 64 && i < 96)) {
+                            eastWires.set(baseNode.getWireIndex());
+                        } else {
+                            assert(i < 64 || (isUltraScale || isVersal && i >= 96 && i < 128));
+                            westWires.set(baseNode.getWireIndex());
+                        }
+                    }
                 }
             } else {
-                assert(baseWireName.matches("CTRL_[EW](_B)?\\d+|INT_NODE_GLOBAL_\\d+(_INT)?_OUT[01]?")
-                        // FIXME:
-                        || series == Series.Versal);
+                assert((isUltraScale || isUltraScalePlus) && baseWireName.matches("CTRL_[EW](_B)?\\d+|INT_NODE_GLOBAL_\\d+(_INT)?_OUT[01]?"));
+            }
+        }
+
+        if (isVersal) {
+            for (TileTypeEnum tte : Arrays.asList(TileTypeEnum.INTF_LOCF_TL_TILE,
+                    TileTypeEnum.INTF_LOCF_BL_TILE,
+                    TileTypeEnum.INTF_ROCF_TL_TILE,
+                    TileTypeEnum.INTF_ROCF_BL_TILE,
+                    TileTypeEnum.INTF_LOCF_TR_TILE,
+                    TileTypeEnum.INTF_LOCF_BR_TILE,
+                    TileTypeEnum.INTF_ROCF_TR_TILE,
+                    TileTypeEnum.INTF_ROCF_BR_TILE)) {
+                Tile intfTile = device.getArbitraryTileOfType(tte);
+                BitSet[] eastWestWires = this.eastWestWires.computeIfAbsent(tte,
+                        k -> new BitSet[]{new BitSet(), new BitSet()});
+                BitSet eastWires = eastWestWires[0];
+                BitSet westWires = eastWestWires[1];
+                for (int wireIndex = 0; wireIndex < intfTile.getWireCount(); wireIndex++) {
+                    IntentCode baseIntentCode = intfTile.getWireIntentCode(wireIndex);
+                    if (baseIntentCode != IntentCode.NODE_INTF_BNODE && baseIntentCode != IntentCode.NODE_INTF_CNODE) {
+                        continue;
+                    }
+
+                    if (EnumSet.of(TileTypeEnum.INTF_LOCF_TR_TILE,
```

Comment:
Also move out to static instantiation?  Looks like its in a double-nested for loop.